src/features/references.zig

```diff
@@ -243,3 +241,169 @@ pub fn symbolReferences(
+                        );
+
+                        const left_type_node = switch (left_type.type.data) {
+                            .other => |n| n,
+                            else => return,
+                        };
+
+                        const child = (try builder.analyser.lookupSymbolContainer(
+                            .{ .node = left_type_node, .handle = left_type.handle },
+                            offsets.tokenToSlice(tree, datas[called_node].rhs),
+                            !left_type.type.is_type_val,
+                        )) orelse return;
+
+                        if (builder.decl_handle.eql(child)) {
+                            try builder.add(handle, node);
+                        }
+                    },
+                    else => {},
+                }
+            },
+            else => {},
+        }
+    }
+};
+
+pub fn callsiteReferences(
+    allocator: std.mem.Allocator,
+    analyser: *Analyser,
+    decl_handle: Analyser.DeclWithHandle,
+    /// add `decl_handle` as a references
+    include_decl: bool,
+    /// exclude references from the std library
+    skip_std_references: bool,
+    /// search other files for references
+    workspace: bool,
+) error{OutOfMemory}!std.ArrayListUnmanaged(Callsite) {
+    std.debug.assert(decl_handle.decl.* == .ast_node);
+
+    var builder = CallBuilder{
+        .allocator = allocator,
+        .analyser = analyser,
+        .decl_handle = decl_handle,
+    };
+    errdefer builder.deinit();
+
+    const curr_handle = decl_handle.handle;
+    if (include_decl) try builder.add(curr_handle, decl_handle.nameToken());
+
+    try builder.collectReferences(curr_handle, 0);
+
+    if (!workspace) return builder.callsites;
+
+    var dependencies = std.StringArrayHashMapUnmanaged(void){};
+    defer {
+        for (dependencies.keys()) |uri| {
+            allocator.free(uri);
+        }
+        dependencies.deinit(allocator);
+    }
+
+    for (analyser.store.handles.values()) |handle| {
+        if (skip_std_references and std.mem.indexOf(u8, handle.uri, "std") != null) {
+            if (!include_decl or !std.mem.eql(u8, handle.uri, curr_handle.uri))
+                continue;
+        }
+
+        var handle_dependencies = std.ArrayListUnmanaged([]const u8){};
+        defer handle_dependencies.deinit(allocator);
+        try analyser.store.collectDependencies(allocator, handle.*, &handle_dependencies);
+
+        try dependencies.ensureUnusedCapacity(allocator, handle_dependencies.items.len);
+        for (handle_dependencies.items) |uri| {
+            var gop = dependencies.getOrPutAssumeCapacity(uri);
+            if (gop.found_existing) {
+                allocator.free(uri);
+            }
+        }
+    }
+
+    for (dependencies.keys()) |uri| {
+        if (std.mem.eql(u8, uri, curr_handle.uri)) continue;
+        const handle = analyser.store.getOrLoadHandle(uri) orelse continue;
+
+        try builder.collectReferences(handle, 0);
+    }
```

Comment:
This looks identical to src/features/references.zig#L183-L215, would it be better to refactor it out?

Reply:
Good call, I'll do that now :)